### PR TITLE
[CMake 2/3] Add CMake Build to torchtext to create single `_torchtext` library

### PR DESCRIPTION
**Pull request by Nayef211**
**Reference Issue #1644**

## Description
- Updating torchtext extension to use CMake instead of PyTorch's `CppExtension` module
- This is the second part of #1660. We create a single `_torchtext.so` library with all source files from csrc/ folder. This is similar to the build artifacts created by the current build process.

## Build Process
**Building the torctext project using `setup.py`**
```shell
python setup.py develop
```

## Misc
closes https://github.com/pytorch/text/pull/1225

### CODE REVIEW

File: torchtext/__init__.py
Comment (mthrok) [inline torchtext/__init__.py:4]: Please check the compatibility of this initialization update on fbcode.

Also I recommend to call this function in `_extension.py` just for the sake of the simplification of `__init__.py`. (delegating the reponsibility to `_extension.py`, which yields a better locality.)

Still, the `_extension` module has to be imported first in `__init__.py`.

cc @parmeet

File: build_tools/setup_helpers/extension.py
Comment (mthrok) [inline build_tools/setup_helpers/extension.py:82]: This is supposed to be handled by TORCH_CXX_FLAGS.

However, https://github.com/pytorch/audio/issues/2287 suggests this does not work with manually built PyTorch, so I will follow up with this one.

File: build_tools/setup_helpers/extension.py
Comment (mthrok) [inline build_tools/setup_helpers/extension.py:80]:
```suggestion
            "-DBUILD_TESTING:BOOL=OFF", "-DBUILD_SHARED_LIBS=OFF",
```

Comment (Nayef211) [reply build_tools/setup_helpers/extension.py:80]: Oops thanks for catching this!

File: build_tools/setup_helpers/extension.py
Comment (Nayef211) [reply build_tools/setup_helpers/extension.py:82]: Gotcha, I will keep this in for now until the issue is resolved then.

File: torchtext/__init__.py
Comment (Nayef211) [reply torchtext/__init__.py:4]: Updated the function call to reside within `_extension.py`. Will import this entire PR and test on fbcode before landing once we get all GH CI tests passing

File: torchtext/_extension.py
Comment (Nayef211) [reply torchtext/_extension.py:59]: Changed this back to raise an `Exception` instead of a `warning`

File: .circleci/unittest/linux/scripts/setup_env.sh
Comment (mthrok) [inline .circleci/unittest/linux/scripts/setup_env.sh:39]: Maybe we need an equivalent change in windows setup script.

It seems like an old version of CMake can cause dynamic link issue

https://stackoverflow.com/a/37146557

File: build_tools/setup_helpers/extension.py
Comment (mthrok) [inline build_tools/setup_helpers/extension.py:114]: This affects flags of all the compilation.

In modern CMake it is recommended to approach this in more OOP-like fashion, using `target`s.

Once we verify this works, then we should move these dedicated location.

File: build_tools/setup_helpers/extension.py
Comment (Nayef211) [reply build_tools/setup_helpers/extension.py:22]: Re-added these flags to the compile commands

File: torchtext/csrc/CMakeLists.txt
Comment (mthrok) [inline torchtext/csrc/CMakeLists.txt:14]: Okay, so probably the correct way is to set this to `STATIC` when in windows.

```
if (MSVC)
    add_library(${name} STATIC ${sources})
else (MSVC)
    add_library(${name} SHARED ${sources})
endif()
```

File: .circleci/unittest/linux/scripts/install.sh
Comment (mthrok) [inline .circleci/unittest/linux/scripts/install.sh:28]: `CONDA_CHANNEL_FLAGS` and `MKL_CONSTANT` can be empty based on platforms, so the use of double-quotations here cause problem. The `command` will see it as option of empty string and it causes weird behavior.
```suggestion
    conda install -y -c "pytorch-${UPLOAD_CHANNEL}" ${CONDA_CHANNEL_FLAGS} ${MKL_CONSTRAINT} pytorch cpuonly
```

Comment (mthrok) [reply .circleci/unittest/linux/scripts/install.sh:28]: BTW why does it have `cpuonly`? It does not hurt but it's not mandatory, is it? TorchText should build fine with non-CPU version of PyTorch (at least CUDA).

Comment (Nayef211) [reply .circleci/unittest/linux/scripts/install.sh:28]: > CONDA_CHANNEL_FLAGS and MKL_CONSTANT can be empty based on platforms, so the use of double-quotations here cause problem. The command will see it as option of empty string and it causes weird behavior.

Gotcha will change this back to remove the double quotes

> BTW why does it have cpuonly? It does not hurt but it's not mandatory, is it? TorchText should build fine with non-CPU version of PyTorch (at least CUDA).

Not sure what the historical context was for this. But also torchtext doesn't need the CUDA version of PyTorch to run. So does it make sense to install the CUDA version when it's not necessary for our library? In addition the CUDA version of PyTorch is a larger than the CPU only version.

File: CMakeLists.txt
Comment (mthrok) [inline CMakeLists.txt:52]: note: Technically, this one is not necessary as there is no `cmake` directory which houses custom CMake scripts.
But it does not hurt to keep this as when such thing is added, it's customary to use `cmake` as directory name.

File: torchtext/__init__.py
Comment (mthrok) [inline torchtext/__init__.py:3]: Can you add a note that this is done for side-effect and it has to happen before anything?

File: torchtext/__init__.py
Comment (mthrok) [inline torchtext/__init__.py:1]: off-topic, but it looks strange that the top-level module initialization script, which should be only doing initializations require os and do some stuff. These should be done in a dedicated submodule and only imported here.

File: torchtext/_extension.py
Comment (mthrok) [inline torchtext/_extension.py:51]: FYI: The implementation beneath these two different `load_library` are common, so only one of them are required.

File: torchtext/csrc/CMakeLists.txt
Comment (mthrok) [inline torchtext/csrc/CMakeLists.txt:43]: since all the headers are now looked up from the root directory of the repo, so I think it's less confusing not to have headers listed here.

**Review by mthrok — CHANGES_REQUESTED** [review]
Overall looks good. Thanks for making this work!

File: build_tools/setup_helpers/extension.py
Comment (Nayef211) [reply build_tools/setup_helpers/extension.py]: Updated to use `target`s. Thanks for the feedback here and I'm glad we figured out that switching to this approach resolved the Windows errors.

File: torchtext/__init__.py
Comment (Nayef211) [reply torchtext/__init__.py:3]: Done

File: torchtext/__init__.py
Comment (Nayef211) [reply torchtext/__init__.py:1]: Let me do this in a followup PR. Maybe we can define all global constants in a `_constants.py` file. Wdyt @parmeet?

File: torchtext/_extension.py
Comment (Nayef211) [reply torchtext/_extension.py:51]: Hmm so I noticed audio actually makes a call to both. Any specific reason why it's done that way in torchaudio?

https://github.com/pytorch/audio/blob/main/torchaudio/_extension.py#L51-L52

File: torchtext/csrc/CMakeLists.txt
Comment (Nayef211) [reply torchtext/csrc/CMakeLists.txt]: Done!

File: .circleci/unittest/linux/scripts/install.sh
Comment (mthrok) [inline .circleci/unittest/linux/scripts/install.sh:28]:
```suggestion
    conda install -y -c "pytorch-${UPLOAD_CHANNEL}" ${CONDA_CHANNEL_FLAGS} ${MKL_CONSTRAINT} pytorch cpuonly
```

File: torchtext/_extension.py
Comment (mthrok) [inline torchtext/_extension.py:5]: Is this `F401` necessary??

File: torchtext/_extension.py
Comment (mthrok) [reply torchtext/_extension.py:51]: Yeah, I learned this fact myself only after I wrote this code. I have not updated audio yet.

File: CMakeLists.txt
Comment (Nayef211) [reply CMakeLists.txt:52]: Gotcha will keep that in mind. What sort of scripts would we put in a `cmake/` folder that wouldn't go in this CMake file?

**Review by mthrok — APPROVED** [review]
I thought I put "Approved" in the previous review, but somehow it was "Request changes".

The change looks good, this can be merged when all the builds / unit tests pass.

File: torchtext/_extension.py
Comment (Nayef211) [reply torchtext/_extension.py:51]: Okay let me go with `torch.ops.load_library` since that seems to be the approach vision also uses https://github.com/pytorch/vision/blob/9f48c6274f66f71dd0cabe6a48bde84c8f239abe/torchvision/extension.py#L20

File: torchtext/_extension.py
Comment (Nayef211) [reply torchtext/_extension.py:5]: You're right it's not since `_mod_utils` is used in the same file.

**Comment by Nayef211** [comment]
@parmeet I'm seeing some test failures on this PR that have nothing to do with my changes. Do you know if it would be fine to land this?

**Pull request merged**
